### PR TITLE
CNFSFile: retry if nfs_open returns EAGAIN

### LINKED ISSUE #22566: NFS4ERR_EXPIRED(-11)
### Describe the bug
Here is a clear and concise description of what the problem is:

I'm not sure what the pattern is, but sometimes seemingly randomly I can't play files from my NFS4 source.
The error in the log file is `open call failed with "NFS4: (path [...]) failed with NFS4ERR_EXPIRED(-11)"`.

This persists until I restart Kodi, after which it works fine. NFS4 is pretty new so I may discover what the pattern is later, only been using it for a week or so.

## Possible Fix
Restarting Kodi fixes it, but that's not ideal.

### To Reproduce
Unsure, this seems random so far.

### Debuglog
We've got some works going in the house and most of my stuff is off and hard to access, so I don't have an easy way now to grab the log file off my nvidia shield.
If the above error message is not enough I'll have to get that log file later on, but I'm assuming I'm not the only one affected.

### Your Environment
Used Operating system:

 - [X] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Android TV 11 (nvidia shield pro)
 - Kodi version: 20

The NFS4 server is unRAID, so nothing special there it's just Linux.
I am using the Jellyfin addon in native mode, my understanding is that it doesn't actively do anything with storage and just relies on Kodi there, but just putting it there in case it matters.

Thanks

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by lrusak**
That means the nfsv4 context has expired and needs to be recreated

fixes #22566

This moves the `nfs_open` method into a do/while loop that will iterate while the return value is less than zero and we still have retry attempts remaining. I've limited it to 1 retry.

NFSv4 has a lease time which needs which means if the don't have a connection in the timeout period we will receive `NFS4ERR_EXPIRED` and `EAGAIN` in which case we need to recreate the nfs context.

We could probably adjust the keep alive parameters but nfsd allows setting a different lease time so the keep alive would have to be configurable as well. It looks like the default lease time for nfsd is 90 seconds.

ref: https://man7.org/linux/man-pages/man8/nfsd.8.html
ref: https://git.linux-nfs.org/?p=steved/nfs-utils.git;a=blob;f=nfs.conf;h=323f072b2fc0a92539d3c9f10620fee94b806f5a;hb=HEAD#l72

You will now see something like this:
```
2023-02-06 14:45:38.721 T:1062417   debug <general>: CFileCache::Open - <nfs://192.168.1.108/mnt/tank/movies/Transformers Dark of the Moon (2011)/Transformers Dark of the Moon.mkv> opening
2023-02-06 14:45:38.721 T:1062417   debug <general>: NFS: Context for 192.168.1.108/ not open - get a new context.
2023-02-06 14:45:38.721 T:1062417   debug <general>: NFS: version: 4
2023-02-06 14:45:38.727 T:1062417   debug <general>: NFS: Connected to server 192.168.1.108 and export /
2023-02-06 14:45:38.727 T:1062417   debug <general>: NFS: chunks: r/w 0/0
2023-02-06 14:45:38.743 T:1062417   debug <general>: CNFSFile::Open - opened mnt/tank/movies/Transformers Dark of the Moon (2011)/Transformers Dark of the Moon.mkv
2023-02-06 14:45:38.743 T:1062417   debug <general>: CFileCache::Open - <nfs://192.168.1.108/mnt/tank/movies/Transformers Dark of the Moon (2011)/Transformers Dark of the Moon.mkv> source chunk size is 0, setting cache chunk size to 131072
2023-02-06 14:45:38.743 T:1062417   debug <general>: CFileCache::Open - <nfs://192.168.1.108/mnt/tank/movies/Transformers Dark of the Moon (2011)/Transformers Dark of the Moon.mkv> using double memory cache each sized 10485760 bytes
```

I did also test this with NFSv3 and it seemed fine.

### CODE REVIEW

**Review by arnova — APPROVED** [review]

**Comment by lrusak** [comment]
I've added a fixup to use a lambda instead of the do/while loop. I think this looks cleaner as we only retry once anyways. @graysky2 can you try again?

**Comment by arnova** [comment]
> I've added a fixup to use a lambda instead of the do/while loop. I think this looks cleaner as we only retry once anyways. @graysky2 can you try again as it will now retry on all errors.
> 
> ```
> 2023-02-07 20:25:59.547 T:1198318   debug <general>: CFileCache::Open - <nfs://192.168.1.108/mnt/tank/movies/Star Trek Into Darkness (2013)/Star Trek Into Darkness.mkv> opening
> 2023-02-07 20:25:59.548 T:1198318   debug <general>: CNFSFile::Open: Unable to open file : 'mnt/tank/movies/Star Trek Into Darkness (2013)/Star Trek Into Darkness.mkv' errno: Resource temporarily unavailable (11) error : 'open call failed with "NFS4: (path /mnt/tank/movies/Star Trek Into Darkness (2013)) failed with NFS4ERR_EXPIRED(-5)"'
> 2023-02-07 20:25:59.548 T:1198318   debug <general>: CNFSFile::Open: Trying again with a new context
> 2023-02-07 20:25:59.548 T:1198318   debug <general>: NFS: Context for 192.168.1.108/ not open - get a new context.
> 2023-02-07 20:25:59.548 T:1198318   debug <general>: NFS: version: 4
> 2023-02-07 20:25:59.552 T:1198318   debug <general>: NFS: Connected to server 192.168.1.108 and export /
> 2023-02-07 20:25:59.553 T:1198318   debug <general>: NFS: chunks: r/w 0/0
> 2023-02-07 20:25:59.566 T:1198318   debug <general>: CNFSFile::Open - opened mnt/tank/movies/Star Trek Into Darkness (2013)/Star Trek Into Darkness.mkv
> ```

Perhaps make the lambda just a (private) class method? Maybe even cleaner?

**Comment by graysky2** [comment]
@lrusak - I have been running a build with the updated version of this PR with the same NFSv4 exports for several hours now and I have not experienced any errors.  Good job :+1:

**Comment by graysky2** [comment]
Still working as expected after multiple days of uptime (cherry picked against latest from nexus branch).

File: xbmc/filesystem/NFSFile.cpp [hunk 1]
```diff
@@ -586,21 +585,37 @@ bool CNFSFile::Open(const CURL& url)
 
   std::unique_lock<CCriticalSection> lock(gNfsConnection);
 
-  if(!gNfsConnection.Connect(url, filename))
-    return false;
+  auto NfsOpen = [this](const CURL& url, std::string& filename) -> int
```
Comment (enen92) [inline xbmc/filesystem/NFSFile.cpp:588]: You seem to return bools in different locations (implicit conversion to int) and don't seem to use the return code of nfs_open. Why not make the lambda return bool right away?

**Comment by lrusak** [comment]
updated
```
2023-02-11 19:58:05.572 T:1685898   debug <general>: connect replacing configured host 192.168.1.108 with resolved host 192.168.1.108
2023-02-11 19:58:05.578 T:1686246   debug <general>: CFileCache::Open - <nfs://192.168.1.108/mnt/tank/movies/Transformers Dark of the Moon (2011)/Transformers Dark of the Moon.mkv> opening
2023-02-11 19:58:05.579 T:1686246   error <general>: CNFSFile::Open: Unable to open file - trying again with a new context: error: 'open call failed with "NFS4: (path /mnt/tank/movies/Transformers Dark of the Moon (2011)) failed with NFS4ERR_EXPIRED(-5)"'
2023-02-11 19:58:05.579 T:1686246   debug <general>: NFS: Context for 192.168.1.108/ not open - get a new context.
2023-02-11 19:58:05.579 T:1686246   debug <general>: NFS: version: 4
2023-02-11 19:58:05.583 T:1686246   debug <general>: NFS: Connected to server 192.168.1.108 and export /
2023-02-11 19:58:05.583 T:1686246   debug <general>: NFS: chunks: r/w 0/0
2023-02-11 19:58:05.600 T:1686246   debug <general>: CNFSFile::Open - opened mnt/tank/movies/Transformers Dark of the Moon (2011)/Transformers Dark of the Moon.mkv

```

File: xbmc/filesystem/NFSFile.cpp (hunk 1 above)
Comment (lrusak) [reply xbmc/filesystem/NFSFile.cpp]: yea, I originally was going to check the ret value, but it seems it's not needed at all. I'll update it to `bool`

File: xbmc/filesystem/NFSFile.cpp [hunk 2]
```diff
@@ -586,21 +585,37 @@ bool CNFSFile::Open(const CURL& url)
...
-              url.GetFileName(), nfs_get_error(m_pNfsContext));
-    m_pNfsContext = NULL;
-    m_exportPath.clear();
-    return false;
+    CLog::Log(LOGDEBUG, "CNFSFile::Open: Unable to open file : '{}' errno: {} ({}) error : '{}'",
+              url.GetFileName(), strerror(errno), errno, nfs_get_error(m_pNfsContext));
+
+    gNfsConnection.Deinit();
+
+    CLog::Log(LOGDEBUG, "CNFSFile::Open: Trying again with a new context");
```
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/filesystem/NFSFile.cpp]: done

**Review by wsnipex — APPROVED** [review]

**Comment by lrusak** [comment]
squashed.

The diff looks bad so I suggest using the split view to see the change. https://github.com/xbmc/xbmc/pull/22714/files?diff=split&w=0

**Comment by fritsch** [comment]
Change looks good. Do you consider backport for Nexus?

While the "newish" C++ is quite hard to read, I find it never the less a bit funny when looking at the mix with globals ;-)

**Review by fritsch — APPROVED** [review]

**Comment by thexai** [comment]
Now crash opening any Blu-Ray BDMV folder structure using NFSv3:

crash

stack

100% reproducible in Android (Shield) and Windows:

```
2023-02-22 10:38:47.222 T:3260    debug <general>: NFS: Connected to server 192.168.50.13 and export /volume1/NAS
2023-02-22 10:38:47.222 T:3260    debug <general>: NFS: chunks: r/w 131072/131072
2023-02-22 10:38:47.223 T:3260    error <general>: CNFSFile::Open: Unable to open file: 'volume1/NAS/TEST-MEDIA/UHD/Full_UHD/Uncharted (2021) UHD/AACS/Unit_Key_RO.inf' error: 'open call failed with "NFS: Lookup of /TEST-MEDIA/UHD/Full_UHD/Uncharted (2021) UHD/AACS/Unit_Key_RO.inf failed with NFS3ERR_NOENT(-2)"'
Exception thrown at 0x00007FF64D8724CB in kodi.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.
```

I think in general the crash will occur every time you try to open a file that doesn't exist....

**Comment by thexai** [comment]
Some things I think are wrong in this PR:

return inside lambda code only returns from lambda and continues:

```
    if (!gNfsConnection.Connect(url, filename))
      return false;
```

Before was return outside main method, now only returns from lambda and continues. Then failure is evaluated same as fail is due `nfs_open` in `nfs_get_error`.

ie: this log may be inaccurate:

```
    CLog::Log(LOGERROR,
              "CNFSFile::Open: Unable to open file - trying again with a new context: error: '{}'",
              nfs_get_error(m_pNfsContext));
```


`nfs_open` returns integer error code and is ignored converted to true/false only.

ie: is not necessary retry if error code is `NFS3ERR_NOENT(-2) `

**Comment by thexai** [comment]
Possible fix:

```c++
bool CNFSFile::Open(const CURL& url)
{
  Close();
  // we can't open files like nfs://file.f or nfs://server/file.f
  // if a file matches the if below return false, it can't exist on a nfs share.
  if (!IsValidFile(url.GetFileName()))
  {
    CLog::Log(LOGINFO, "NFS: Bad URL : '{}'", url.GetFileName());
    return false;
  }

  constexpr int NFS4ERR_EXPIRED = -11;
  constexpr int NFSERR_CONNECTION = 1;

  std::string filename;

  std::unique_lock<CCriticalSection> lock(gNfsConnection);

  auto NfsOpen = [this](const CURL& url, std::string& filename) -> int
  {
    if (!gNfsConnection.Connect(url, filename))
      return NFSERR_CONNECTION;

    m_pNfsContext = gNfsConnection.GetNfsContext();
    m_exportPath = gNfsConnection.GetContextMapId();

    return nfs_open(m_pNfsContext, filename.c_str(), O_RDONLY, &m_pFileHandle);
  };

  int ret = NfsOpen(url, filename);

  if (ret == NFSERR_CONNECTION)
  {
    return false;
  }
  else if (ret == NFS4ERR_EXPIRED)
  {
    CLog::Log(LOGWARNING,
              "CNFSFile::Open: Unable to open file - trying again with a new context: error: '{}'",
              nfs_get_error(m_pNfsContext));

    gNfsConnection.Deinit();
    ret = NfsOpen(url, filename);
  }

  if (ret != 0)
  {
    CLog::Log(LOGERROR, "CNFSFile::Open: Unable to open file: '{}' error: '{}'", url.GetFileName(),
              nfs_get_error(m_pNfsContext));

    m_pNfsContext = nullptr;
    m_exportPath.clear();
    return false;
  }

  CLog::Log(LOGDEBUG, "CNFSFile::Open - opened {}", url.GetFileName());
  m_url=url;

  struct __stat64 tmpBuffer;

  if( Stat(&tmpBuffer) )
  {
    m_url.Reset();
    Close();
    return false;
  }

  m_fileSize = tmpBuffer.st_size;//cache the size of this file
  // We've successfully opened the file!
  return true;
}
```

**Pull request merged**